lib/Serialization/Deserialization.cpp

```diff
@@ -6472,6 +6473,43 @@ llvm::Error DeclDeserializer::deserializeDeclCommon() {
   }
 }
 
+/// Complete attributes that contain recursive references to the decl being
+/// deserialized or to other decls. This method is called after \p decl is
+/// created and stored into the \c ModuleFile::Decls table, so any cycles
+/// between mutually-referencing decls will be broken.
+///
+/// Attributes handled here include:
+///
+///  \li \c \@differentiable
+///  \li \c \@derivative
+///  \li \c \@abi
+void DeclDeserializer::finishRecursiveAttrs(Decl *decl, DeclAttribute *attrs) {
+  DeclAttributes tempAttrs;
+  tempAttrs.setRawAttributeChain(attrs);
+
+  // @differentiable and @derivative
+  for (auto *attr : tempAttrs.getAttributes<DifferentiableAttr>()) {
+    auto *diffAttr = const_cast<DifferentiableAttr *>(attr);
+    diffAttr->setOriginalDeclaration(decl);
+    diffAttr->setParameterIndices(diffAttrParamIndicesMap[diffAttr]);
+  }
+  for (auto *attr : tempAttrs.getAttributes<DerivativeAttr>()) {
+    auto *derAttr = const_cast<DerivativeAttr *>(attr);
+    derAttr->setOriginalDeclaration(decl);
+  }
+
+  // @abi
+  if (unresolvedABIAttr) {
+    unresolvedABIAttr->first->abiDecl = MF.getDecl(unresolvedABIAttr->second);
+    ctx.recordABIAttr(unresolvedABIAttr->first, decl);
+  }
+  if (ABIDeclCounterpartID != 0) {
+    // This decl is the `abiDecl` of an `ABIAttr`. Force the decl that `ABIAttr`
+    // belongs to so that we establish the relationship between them.
+    (void)MF.getDecl(ABIDeclCounterpartID);
```

Comment:
Could we protect the two `MF.getDecl` here against errors by instead using `getDeclChecked`? Any error could either be bubbled up or dropped when safe to do so.

Reply:
Might take a bit of work, but probably.

Reply:
Looks good, thank you!